File: src/aiperf/graph/scheduler.py
# SPDX-FileCopyrightText: Copyright (c) 2025-2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
# SPDX-License-Identifier: Apache-2.0
"""Graph-derived adjacency helper for the async-dataflow executor.

A pure adjacency lookup over the parsed graph's static edges; holds no mutable
per-trace state. Per-trace firing dedup lives on
`_TraceContext.scheduled_node_ids`.
"""

from __future__ import annotations

from collections import defaultdict

import msgspec

from aiperf.dataset.graph.models import (
    END_NODE_ID,
    START_NODE_ID,
    GraphRecord,
    StaticEdge,
)


class Scheduler:
    """Graph-derived adjacency helper for the async-dataflow executor.

    All state is derived from the parsed graph at construction time and is
    immutable across traces (no mutable per-trace fields). Per-trace state
    lives on `_TraceContext.scheduled_node_ids`; this class is shared across
    every trace in a phase.
    """

    __slots__ = (
        "_entry",
        "_start_anchored_succ",
        "_static_pred_edges",
        "_static_succ",
    )

    def __init__(self, graph: GraphRecord) -> None:
        _reject_start_anchored_start_edge(graph.edges)
        static_succ: dict[str, list[str]] = defaultdict(list)
        start_anchored_succ: dict[str, list[str]] = defaultdict(list)
        static_pred_edges: dict[str, list[StaticEdge]] = defaultdict(list)

        for edge in graph.edges:
            if edge.delay_after_predecessor_start_us is not None:
                start_anchored_succ[edge.source].append(edge.target)
            else:
                static_succ[edge.source].append(edge.target)
            static_pred_edges[edge.target].append(edge)

        _reject_unanchored_first_token_edge(static_pred_edges)
        _reject_mixed_anchor_fan_in(static_pred_edges)

        entry: list[str] = []
        seen: set[str] = set()
        for target in static_succ.get(START_NODE_ID, []):
            if target == END_NODE_ID or target in seen:
                continue
            seen.add(target)
            entry.append(target)

        self._static_succ: dict[str, list[str]] = dict(static_succ)
        self._start_anchored_succ: dict[str, list[str]] = dict(start_anchored_succ)
        self._static_pred_edges: dict[str, list[StaticEdge]] = dict(static_pred_edges)
        self._entry: list[str] = entry

    def entry_nodes(self) -> list[str]:
        """Return node ids that should fire at trace start (successors of START).

        END is suppressed.
        """
        return list(self._entry)

    def successors_after(self, node_id: str) -> list[str]:
        """Return successor node ids that should fire after `node_id` completes.

        Static successors via StaticEdge. Start-anchored StaticEdges
        (`delay_after_predecessor_start_us`) are EXCLUDED here; they are
        scheduled at `node_id`'s DISPATCH instead of its completion (see
        `start_anchored_successors`). END is suppressed.
        """
        return [t for t in self._static_succ.get(node_id, []) if t != END_NODE_ID]

    def incoming_static_edges(self, node_id: str) -> list[StaticEdge]:
        """Return StaticEdge objects targeting `node_id`.

        Used by edge-gate computation for `min_start_delay_us` /
        `delay_after_predecessor_us` / `delay_after_predecessor_start_us`.
        """
        return list(self._static_pred_edges.get(node_id, []))

    def start_anchored_successors(self, node_id: str) -> list[str]:
        """Successors wired via start-anchored edges. The executor schedules
        these at `node_id`'s DISPATCH (firing-gate clear), not its completion;
        they are deliberately absent from `successors_after` so a child that
        finishes before its still-running parent is not re-scheduled into the
        cycle guard."""
        return [
            t for t in self._start_anchored_succ.get(node_id, []) if t != END_NODE_ID
        ]


def _reject_start_anchored_start_edge(edges: list[StaticEdge]) -> None:
    """Reject an edge whose start anchor refers to the virtual START node."""
    for edge in edges:
        if (
            edge.source == START_NODE_ID
            and edge.delay_after_predecessor_start_us is not None
        ):
            raise NotImplementedError(
                f"edge {START_NODE_ID!r} -> {edge.target!r}: "
                "delay_after_predecessor_start_us is unsupported on the "
                "virtual START node because START is never dispatched. Use "
                "min_start_delay_us for a trace-entry delay."
            )


def _reject_unanchored_first_token_edge(
    static_pred_edges: dict[str, list[StaticEdge]],
) -> None:
    """Reject a first-token-anchored in-edge that carries no start anchor.

    ``TraceExecutor._apply_firing_delay`` blocks on
    ``ctx.first_token_event(edge.source).wait()`` for every incoming edge
    carrying ``delay_after_predecessor_first_token_us``, and that wait has no
    self-imposed bound (``AIPERF_GRAPH_EXECUTOR_WATCHDOG_TIMEOUT`` is unset by
    default). It is safe today only because the edge ALSO carries
    ``delay_after_predecessor_start_us``, which makes it start-anchored, which
    makes :func:`_reject_mixed_anchor_fan_in` guarantee it is its target's ONLY
    in-edge -- so the target is scheduled solely by
    ``start_anchored_successors(source)`` at the source's DISPATCH. The source
    is therefore already running, and ``_finalize_node``'s ``finally`` will set
    the latch on every exit path.

    Strip the start anchor and that whole chain lapses: the target becomes
    reachable from an unrelated predecessor while the first-token source may
    never run, and the wait deadlocks with no diagnostic. The sole producer
    (``interval_order.apply_start_anchors``) always co-emits both, so this
    gates a shape no shipped lowering can currently emit -- it exists to make a
    future lowering that breaks the pairing fail loudly at graph load instead
    of hanging a benchmark.

    A completion anchor (``delay_after_predecessor_us``) does NOT satisfy this:
    it leaves the edge classified as a completion edge, so the sole-in-edge
    guarantee still never applies.
    """
    for target, edges in static_pred_edges.items():
        for edge in edges:
            if edge.delay_after_predecessor_first_token_us is None:
                continue
            if edge.delay_after_predecessor_start_us is not None:
                continue
            raise NotImplementedError(
                f"node {target!r}: first-token-anchored edge "
                f"{edge.source!r} -> {target!r} carries "
                "delay_after_predecessor_first_token_us without "
                "delay_after_predecessor_start_us, which is unsupported. The "
                "executor waits on the source's first-token latch before "
                "computing this node's firing gate, and that wait is only "
                "guaranteed to be satisfiable because a start anchor forces "
                "the edge to be its target's ONLY in-edge (so the target is "
                "scheduled at the source's dispatch, with the source already "
                "running). Without the start anchor the target can be "
                "scheduled by an unrelated predecessor while the source never "
                "runs, and the wait deadlocks. Emit "
                "delay_after_predecessor_start_us on the same edge."
            )


def _reject_mixed_anchor_fan_in(
    static_pred_edges: dict[str, list[StaticEdge]],
) -> None:
    """Reject a start-anchored in-edge that is not its target's ONLY in-edge.

    The runtime half-supports ANY fan-in involving a start anchor: the start
    anchor schedules the target at its anchor parent's DISPATCH, so the target
    fires without waiting for its other predecessors (a completion
    predecessor's recorded ordering is silently ignored; a second start-anchor
    parent that has not yet dispatched is silently dropped by the firing gate)
    and, when the other predecessor later finishes/dispatches, it re-schedules
    the DONE target into the cycle guard (spurious "cycle detected"). No
    shipped lowering emits either shape (`apply_start_anchors` replaces a
    start-anchored node's WHOLE in-edge set with exactly one edge), so it is
    gated loudly at graph load instead.
    """
    for target, edges in static_pred_edges.items():
        start_anchored = [
            e for e in edges if e.delay_after_predecessor_start_us is not None
        ]
        if not start_anchored or len(edges) == 1:
            continue
        completion = [e for e in edges if e.delay_after_predecessor_start_us is None]
        if completion:
            shape = "mixed-anchor fan-in"
            detail = (
                f"start-anchored edge {start_anchored[0].source!r} -> {target!r} "
                f"(delay_after_predecessor_start_us) and completion edge "
                f"{completion[0].source!r} -> {target!r} arrive at the same node"
            )
        else:
            shape = "multi-start-anchored fan-in"
            detail = (
                f"start-anchored edges {start_anchored[0].source!r} -> {target!r} "
                f"and {start_anchored[1].source!r} -> {target!r} "
                f"(delay_after_predecessor_start_us) arrive at the same node"
            )
        raise NotImplementedError(
            f"node {target!r}: {shape} is unsupported: {detail}. "
            "The runtime schedules a start-anchored target at its anchor "
            "parent's DISPATCH, so the node would fire without waiting for "
            "its other predecessor and then be re-scheduled into the cycle "
            "guard when that predecessor finishes or dispatches. A "
            "start-anchored in-edge must be its target's ONLY in-edge."
        )


def collapse_leading_start_offsets(graph: GraphRecord) -> GraphRecord:
    """Zero the leading phase-start offsets (`--burst-phase-starts` collapse).

    The anchor carrier for a firing's t*-relative leading offset is the
    `min_start_delay_us` on its START in-edge: the trie build roots gap-started
    chains at START with the warped arrival offset
    (`interval_order.build_interval_edges`) and the t* snapshot chop re-roots
    every surviving frontier node at START with its `arrival - t*` offset
    (`snapshot_chop._chop_edges`). The executor gates on that edge field
    (`_compute_firing_gate_us`). Burst collapses ONLY those leading offsets --
    every node fires the instant its inputs are ready -- while non-START edges
    keep their recorded inter-turn pacing untouched. Node-level
    `min_start_delay_us` is collapsed too, but ONLY on a node with no non-START
    static in-edge (same leading-anchor semantics): a
    node-level delay on a node that HAS a real predecessor is mid-graph pacing,
    not a leading offset, and must survive burst.

    Pure function: returns a rebuilt `GraphRecord` (msgspec replace), never
    mutates. Identity-preserving for untouched nodes/edges.
    """
    has_real_pred = {
        edge.target for edge in graph.edges if edge.source != START_NODE_ID
    }
    new_edges = [
        msgspec.structs.replace(edge, min_start_delay_us=0.0)
        if edge.source == START_NODE_ID and edge.min_start_delay_us
        else edge
        for edge in graph.edges
    ]
    new_nodes = {
        nid: (
            msgspec.structs.replace(node, min_start_delay_us=0.0)
            if node.min_start_delay_us and nid not in has_real_pred
            else node
        )
        for nid, node in graph.nodes.items()
    }
    return msgspec.structs.replace(graph, nodes=new_nodes, edges=new_edges)

File: rust/runtime/src/graph/scheduler.rs
// SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
// SPDX-License-Identifier: Apache-2.0
//! Graph-derived adjacency helper for the async-dataflow executor.
//!
//! A pure adjacency view over the parsed graph's static edges; holds no per-trace
//! state (that lives on `TraceContext`), so it is shared across every trace.

use crate::graph::model::{END_NODE_ID, GraphRecord, START_NODE_ID, StaticEdge};
use std::collections::BTreeMap;

/// Why an anchored edge topology is unsupported and rejected at construction:
/// a START edge anchored on a dispatch or first-token event it cannot have, or a
/// start-anchored in-edge that is not its target's only in-edge (mixed-anchor /
/// multi-start-anchored fan-in).
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub(crate) enum AnchorFanInKind {
    /// START cannot provide a dispatch or first-token event.
    NonCompletionStart,
    /// A target has both start-anchored and completion-anchored inputs.
    Mixed,
    /// A target has more than one start-anchored input.
    MultipleStartAnchored,
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub(crate) struct AnchorFanInFinding {
    kind: AnchorFanInKind,
    target: String,
    message: String,
}

impl AnchorFanInFinding {
    pub(crate) fn kind(&self) -> AnchorFanInKind {
        self.kind
    }

    pub(crate) fn target(&self) -> &str {
        &self.target
    }

    pub(crate) fn message(&self) -> &str {
        &self.message
    }
}

/// A scheduler construction error for an unsupported anchored fan-in topology.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct MixedAnchorFanInError(pub String);

impl std::fmt::Display for MixedAnchorFanInError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.write_str(&self.0)
    }
}
impl std::error::Error for MixedAnchorFanInError {}

pub struct Scheduler {
    static_succ: BTreeMap<String, Vec<String>>,
    start_anchored_succ: BTreeMap<String, Vec<String>>,
    static_pred_edges: BTreeMap<String, Vec<StaticEdge>>,
    entry: Vec<String>,
}

impl Scheduler {
    pub fn new(graph: &GraphRecord) -> Result<Self, MixedAnchorFanInError> {
        let mut static_succ: BTreeMap<String, Vec<String>> = BTreeMap::new();
        let mut start_anchored_succ: BTreeMap<String, Vec<String>> = BTreeMap::new();
        let mut static_pred_edges: BTreeMap<String, Vec<StaticEdge>> = BTreeMap::new();

        for edge in &graph.edges {
            if let Some(finding) = non_completion_start_finding(edge) {
                return Err(MixedAnchorFanInError(finding.message().to_owned()));
            }
            if edge.delay_after_predecessor_start_us.is_some() {
                start_anchored_succ
                    .entry(edge.source.clone())
                    .or_default()
                    .push(edge.target.clone());
            } else {
                static_succ
                    .entry(edge.source.clone())
                    .or_default()
                    .push(edge.target.clone());
            }
            static_pred_edges
                .entry(edge.target.clone())
                .or_default()
                .push(edge.clone());
        }
        if let Some(finding) = anchor_fan_in_finding_from_predecessors(&static_pred_edges) {
            return Err(MixedAnchorFanInError(finding.message().to_owned()));
        }

        // Entry nodes: successors of START (dedup, END suppressed), edge order.
        let mut entry: Vec<String> = Vec::new();
        let mut seen: std::collections::HashSet<String> = std::collections::HashSet::new();
        if let Some(targets) = static_succ.get(START_NODE_ID) {
            for target in targets {
                if target == END_NODE_ID || seen.contains(target) {
                    continue;
                }
                seen.insert(target.clone());
                entry.push(target.clone());
            }
        }

        Ok(Scheduler {
            static_succ,
            start_anchored_succ,
            static_pred_edges,
            entry,
        })
    }

    /// Node ids that fire at trace start (successors of START; END suppressed).
    ///
    /// Borrows the precomputed entry list; hot scheduling path, so no per-call
    /// allocation or id clone.
    pub fn entry_nodes(&self) -> impl Iterator<Item = &str> {
        self.entry.iter().map(String::as_str)
    }

    /// Static successors after `node_id` completes (start-anchored excluded; END
    /// suppressed).
    ///
    /// Borrows the immutable adjacency; hot scheduling path, so no per-call
    /// allocation or id clone.
    pub fn successors_after(&self, node_id: &str) -> impl Iterator<Item = &str> {
        self.static_succ
            .get(node_id)
            .into_iter()
            .flatten()
            .filter(|t| *t != END_NODE_ID)
            .map(String::as_str)
    }

    /// StaticEdge objects targeting `node_id` (for gate computation).
    pub fn incoming_static_edges(&self, node_id: &str) -> &[StaticEdge] {
        self.static_pred_edges
            .get(node_id)
            .map(Vec::as_slice)
            .unwrap_or(&[])
    }

    /// Successors wired via start-anchored edges; scheduled at `node_id`'s
    /// DISPATCH, not completion. END suppressed.
    ///
    /// Borrows the immutable adjacency; hot scheduling path, so no per-call
    /// allocation or id clone.
    pub fn start_anchored_successors(&self, node_id: &str) -> impl Iterator<Item = &str> {
        self.start_anchored_succ
            .get(node_id)
            .into_iter()
            .flatten()
            .filter(|t| *t != END_NODE_ID)
            .map(String::as_str)
    }
}

pub(crate) fn anchor_fan_in_finding(graph: &GraphRecord) -> Option<AnchorFanInFinding> {
    let mut static_pred_edges: BTreeMap<String, Vec<StaticEdge>> = BTreeMap::new();
    for edge in &graph.edges {
        if let Some(finding) = non_completion_start_finding(edge) {
            return Some(finding);
        }
        static_pred_edges
            .entry(edge.target.clone())
            .or_default()
            .push(edge.clone());
    }
    anchor_fan_in_finding_from_predecessors(&static_pred_edges)
}

fn non_completion_start_finding(edge: &StaticEdge) -> Option<AnchorFanInFinding> {
    (edge.source == START_NODE_ID
        && (edge.delay_after_predecessor_start_us.is_some()
            || edge.delay_after_predecessor_first_token_us.is_some()))
    .then(|| AnchorFanInFinding {
        kind: AnchorFanInKind::NonCompletionStart,
        target: edge.target.clone(),
        message: format!(
            "START edge to {:?} must use completion anchoring because START has no dispatch or first-token event",
            edge.target
        ),
    })
}

fn anchor_fan_in_finding_from_predecessors(
    static_pred_edges: &BTreeMap<String, Vec<StaticEdge>>,
) -> Option<AnchorFanInFinding> {
    for (target, edges) in static_pred_edges {
        let start_anchored: Vec<&StaticEdge> = edges
            .iter()
            .filter(|e| e.delay_after_predecessor_start_us.is_some())
            .collect();
        if start_anchored.is_empty() || edges.len() == 1 {
            continue;
        }
        let completion: Vec<&StaticEdge> = edges
            .iter()
            .filter(|e| e.delay_after_predecessor_start_us.is_none())
            .collect();
        let (kind, shape, detail) = if let Some(comp) = completion.first() {
            (
                AnchorFanInKind::Mixed,
                "mixed-anchor fan-in",
                format!(
                    "start-anchored edge {:?} -> {:?} (delay_after_predecessor_start_us) and \
completion edge {:?} -> {:?} arrive at the same node",
                    start_anchored[0].source, target, comp.source, target
                ),
            )
        } else {
            (
                AnchorFanInKind::MultipleStartAnchored,
                "multi-start-anchored fan-in",
                format!(
                    "start-anchored edges {:?} -> {:?} and {:?} -> {:?} \
(delay_after_predecessor_start_us) arrive at the same node",
                    start_anchored[0].source, target, start_anchored[1].source, target
                ),
            )
        };
        return Some(AnchorFanInFinding {
            kind,
            target: target.clone(),
            message: format!(
                "node {target:?}: {shape} is unsupported: {detail}. A start-anchored in-edge \
must be its target's ONLY in-edge."
            ),
        });
    }
    None
}

/// Zero the leading phase-start offsets (`--burst-phase-starts` collapse).
///
/// Returns a rebuilt graph; identity-preserving for untouched nodes/edges.
pub fn collapse_leading_start_offsets(graph: &GraphRecord) -> GraphRecord {
    let has_real_pred: std::collections::HashSet<&str> = graph
        .edges
        .iter()
        .filter(|e| e.source != START_NODE_ID)
        .map(|e| e.target.as_str())
        .collect();

    let mut new_graph = graph.clone();
    for edge in &mut new_graph.edges {
        if edge.source == START_NODE_ID && edge.min_start_delay_us.unwrap_or(0.0) != 0.0 {
            edge.min_start_delay_us = Some(0.0);
        }
    }
    for (nid, node) in new_graph.nodes.iter_mut() {
        if let Some(node) = node.as_llm_mut() {
            let has_delay = node.min_start_delay_us.unwrap_or(0.0) != 0.0;
            if has_delay && !has_real_pred.contains(nid.as_str()) {
                node.min_start_delay_us = Some(0.0);
            }
        }
    }
    new_graph
}

#[cfg(test)]
mod tests {
    use super::*;

    fn graph(json: &str) -> GraphRecord {
        serde_json::from_str(json).unwrap()
    }

    #[test]
    fn public_error_tuple_source_compatibility() {
        let fan_in = MixedAnchorFanInError("fan-in".to_owned());
        assert_eq!(fan_in.0, "fan-in");
    }

    #[test]
    fn error_compatibility_classifier_reads_existing_predecessor_adjacency() {
        let g = graph(
            r#"{
            "nodes": {"a": {"node_type":"llm","prompt":[],"output":"oa"},
                      "b": {"node_type":"llm","prompt":[],"output":"ob"},
                      "c": {"node_type":"llm","prompt":[],"output":"oc"}},
            "edges": [
                {"edge_type":"static","source":"a","target":"c","delay_after_predecessor_start_us":5.0},
                {"edge_type":"static","source":"b","target":"c"}
            ]
        }"#,
        );
        let mut predecessors = BTreeMap::new();
        for edge in &g.edges {
            predecessors
                .entry(edge.target.clone())
                .or_insert_with(Vec::new)
                .push(edge.clone());
        }

        let finding =
            anchor_fan_in_finding_from_predecessors(&predecessors).expect("mixed fan-in finding");
        assert_eq!(finding.kind(), AnchorFanInKind::Mixed);
        assert_eq!(finding.target(), "c");
    }

    #[test]
    fn entry_successors_and_end_suppressed() {
        let g = graph(
            r#"{
            "nodes": {
                "a": {"node_type":"llm","prompt":[],"output":"oa"},
                "b": {"node_type":"llm","prompt":[],"output":"ob"}
            },
            "edges": [
                {"edge_type":"static","source":"START","target":"a"},
                {"edge_type":"static","source":"a","target":"b"},
                {"edge_type":"static","source":"b","target":"END"}
            ]
        }"#,
        );
        let sched = Scheduler::new(&g).unwrap();
        assert_eq!(sched.entry_nodes().collect::<Vec<_>>(), vec!["a"]);
        assert_eq!(sched.successors_after("a").collect::<Vec<_>>(), vec!["b"]);
        assert_eq!(sched.successors_after("b").count(), 0); // END suppressed
    }

    #[test]
    fn start_anchored_routed_separately() {
        let g = graph(
            r#"{
            "nodes": {"a": {"node_type":"llm","prompt":[],"output":"oa"},
                      "b": {"node_type":"llm","prompt":[],"output":"ob"}},
            "edges": [
                {"edge_type":"static","source":"START","target":"a"},
                {"edge_type":"static","source":"a","target":"b","delay_after_predecessor_start_us":5.0}
            ]
        }"#,
        );
        let sched = Scheduler::new(&g).unwrap();
        assert_eq!(sched.successors_after("a").count(), 0);
        assert_eq!(
            sched.start_anchored_successors("a").collect::<Vec<_>>(),
            vec!["b"]
        );
    }

    #[test]
    fn start_edges_reject_dispatch_and_first_token_anchors() {
        for timing in [
            "\"delay_after_predecessor_start_us\":5.0",
            "\"delay_after_predecessor_first_token_us\":5.0",
        ] {
            let g = graph(&format!(
                r#"{{"nodes":{{"a":{{"node_type":"llm","prompt":[],"output":"oa"}}}},"edges":[{{"edge_type":"static","source":"START","target":"a",{timing}}}]}}"#
            ));
            assert!(Scheduler::new(&g).is_err(), "{timing}");
        }
    }

    #[test]
    fn mixed_anchor_fan_in_error_compatibility_projects_typed_finding() {
        let g = graph(
            r#"{
            "nodes": {"a": {"node_type":"llm","prompt":[],"output":"oa"},
                      "b": {"node_type":"llm","prompt":[],"output":"ob"},
                      "c": {"node_type":"llm","prompt":[],"output":"oc"}},
            "edges": [
                {"edge_type":"static","source":"a","target":"c","delay_after_predecessor_start_us":5.0},
                {"edge_type":"static","source":"b","target":"c"}
            ]
        }"#,
        );
        let error = match Scheduler::new(&g) {
            Err(error) => error,
            Ok(_) => panic!("mixed-anchor fan-in must be rejected"),
        };
        let finding = anchor_fan_in_finding(&g).expect("mixed fan-in finding");
        assert_eq!(finding.kind(), AnchorFanInKind::Mixed);
        assert_eq!(finding.target(), "c");
        assert_eq!(error.0, finding.message());
    }

    #[test]
    fn multi_start_anchor_fan_in_error_compatibility_projects_typed_finding() {
        let g = graph(
            r#"{
            "nodes": {"a": {"node_type":"llm","prompt":[],"output":"oa"},
                      "b": {"node_type":"llm","prompt":[],"output":"ob"},
                      "c": {"node_type":"llm","prompt":[],"output":"oc"}},
            "edges": [
                {"edge_type":"static","source":"a","target":"c","delay_after_predecessor_start_us":5.0},
                {"edge_type":"static","source":"b","target":"c","delay_after_predecessor_start_us":6.0}
            ]
        }"#,
        );
        let error = match Scheduler::new(&g) {
            Err(error) => error,
            Ok(_) => panic!("multi-start-anchored fan-in must be rejected"),
        };
        let finding = anchor_fan_in_finding(&g).expect("multi-start fan-in finding");
        assert_eq!(finding.kind(), AnchorFanInKind::MultipleStartAnchored);
        assert_eq!(finding.target(), "c");
        assert_eq!(error.0, finding.message());
    }

    #[test]
    fn collapse_zeros_leading_start_offset() {
        let g = graph(
            r#"{
            "nodes": {"a": {"node_type":"llm","prompt":[],"output":"oa","min_start_delay_us":9.0}},
            "edges": [
                {"edge_type":"static","source":"START","target":"a","min_start_delay_us":7.0}
            ]
        }"#,
        );
        let collapsed = collapse_leading_start_offsets(&g);
        assert_eq!(collapsed.edges[0].min_start_delay_us, Some(0.0));
        assert_eq!(
            collapsed
                .nodes
                .get("a")
                .unwrap()
                .as_llm()
                .unwrap()
                .min_start_delay_us,
            Some(0.0)
        );
    }
}
